beluga-aws-sdk/src/jobs/mod.rs

```diff
@@ -0,0 +1,306 @@
+use std::collections::HashMap;
+
+use beluga_mqtt::{MqttClient, QoS, Subscriber};
+use chrono::prelude::Utc;
+use futures::prelude::stream::{StreamExt, TryStreamExt};
+
+pub use self::data::JobStatus;
+use self::data::{
+    GetPendingJobsExecutionReq, GetPendingJobsExecutionResp, JobExecution, JobExecutionSummary,
+    StartNextPendingJobExecutionReq, StartNextPendingJobExecutionResp, UpdateJobExecutionReq,
+    UpdateJobExecutionResp,
+};
+use crate::error::Error;
+use crate::Result;
+
+mod data;
+mod datetime;
+
+#[derive(Debug, Clone)]
+pub struct JobsClient {
+    mqtt: MqttClient,
+    subscriber: Subscriber,
+}
+
+impl JobsClient {
+    pub async fn new(mqtt: MqttClient) -> Result<Self> {
+        let subscriber = mqtt
+            .subscribe_many(
+                [
+                    get_accepted(mqtt.thing_name()).as_str(),
+                    get_rejected(mqtt.thing_name()).as_str(),
+                    start_next_accepted(mqtt.thing_name()).as_str(),
+                    start_next_rejected(mqtt.thing_name()).as_str(),
+                ],
+                QoS::AtLeastOnce,
+            )
+            .await?;
+
+        Ok(Self { mqtt, subscriber })
+    }
+
+    pub async fn get(&mut self) -> Result<(Vec<Job>, Vec<Job>)> {
+        self.mqtt
+            .publish(
+                format!("$aws/things/{}/jobs/get", self.mqtt.thing_name()).as_str(),
+                QoS::AtLeastOnce,
+                false,
+                bytes::Bytes::copy_from_slice(&serde_json::to_vec(&GetPendingJobsExecutionReq {
+                    token: format!("get-jobs-{}", self.mqtt.thing_name()),
+                })?),
+            )
+            .await?;
+
+        let packet = self.subscriber.recv().await?;
+
+        if packet.topic == get_accepted(self.mqtt.thing_name()) {
+            let GetPendingJobsExecutionResp {
+                in_progress_jobs,
+                queued_jobs,
+                token,
+                ..
+            } = serde_json::from_slice::<GetPendingJobsExecutionResp>(&packet.payload)?;
+            if token.is_some_and(|token| token == format!("get-jobs-{}", self.mqtt.thing_name())) {
+                let in_progress = futures::stream::iter(in_progress_jobs.into_iter())
+                    .map(|summary| async { Job::from_summary(summary, self.mqtt.clone()).await })
+                    .buffer_unordered(10)
+                    .try_collect::<Vec<_>>()
+                    .await?;
+
+                let queued = futures::stream::iter(queued_jobs.into_iter())
+                    .map(|summary| async { Job::from_summary(summary, self.mqtt.clone()).await })
+                    .buffer_unordered(10)
```

Comment:
I think it might be an issue if we pre-emptively subscribe to all topics related to every queued job. Let's say the pending queue contains like 200 pending jobs for instance.

I think we could either:

1. return a different "lazy job" type that needs to be upgrade into the real `Job` type.
2. return streams from this function rather than collecting the jobs into vecs eagerly.

What are your thoughts?

Reply:
Hmmm, let's discuss that.